Application/ControlMsrE2/ControlMsrE2.c

```diff
@@ -0,0 +1,242 @@
+        
+        if (ListHeaders[ListHeaderIndex] != NULL) {
+            
+            if (IS_VERBOSE()) {
+                Print (L"Package List: ");
+                PrintGuid (&ListHeaders[ListHeaderIndex]->PackageListGuid);
+                Print (L"\n");
+            }
+            
+            // First package in list
+            EFI_HII_PACKAGE_HEADER* PkgHeader = PADD(ListHeaders[ListHeaderIndex], sizeof(EFI_HII_PACKAGE_LIST_HEADER));
+
+            // For each package in list
+            while (ContextsCount < CONTEXTS_MAX) {
+
+                if (IS_VERBOSE()) {
+                    Print (L"Package Type: %02X ", PkgHeader->Type);
+                }
+                
+                if (PkgHeader->Type == EFI_HII_PACKAGE_END) {
+                    break;
+                }
+                else if (PkgHeader->Type == EFI_HII_PACKAGE_FORMS) {
+                    
+                    EFI_IFR_OP_HEADER* IfrHeader = PADD(PkgHeader, sizeof(EFI_HII_PACKAGE_HEADER));
+                    
+                    if (IfrHeader->OpCode == EFI_IFR_FORM_SET_OP) { // Form Definition must start with FORM_SET_OP
+                        
+                        // Print some Info
+                        if (IS_VERBOSE()) {
+                            Print (L"Form: ");
+                            PrintGuid ((void*) &(((EFI_IFR_FORM_SET*) IfrHeader)->Guid));
+                        }
+                        
+                        if (IfrHeader->Length >= 16 + sizeof(EFI_IFR_FORM_SET)) {
+                            if (IS_VERBOSE()) {
+                                Print (L" Class Guid: ");
+                                PrintGuid (PADD(IfrHeader, sizeof(EFI_IFR_FORM_SET)));
+                                Print(L"\n");
+                            }
+                            
+                            // Checkup for Setup Form
+                            if (CompareGuid (&gEfiHiiPlatformSetupFormsetGuid, PADD(IfrHeader, sizeof(EFI_IFR_FORM_SET)))) {
+                                
+                                Contexts[ContextsCount].searchText = SearchString;
+                                Contexts[ContextsCount].efiHandle = *h;
+                                Contexts[ContextsCount].listHeader = ListHeaders[ListHeaderIndex];
+                                Contexts[ContextsCount].pkgHeader = PkgHeader;
+                                Contexts[ContextsCount].firstIfrHeader = PADD(IfrHeader, IfrHeader->Length);
+                                Contexts[ContextsCount].ifrVarStore = NULL;
+                                Contexts[ContextsCount].ifrOneOf = NULL;
+                                Contexts[ContextsCount].stopAt = DONT_STOP_AT;
+                                Contexts[ContextsCount].count = OptionsCount;
+                                
+                                DoForEachOpCode (Contexts[ContextsCount].firstIfrHeader, EFI_IFR_ONE_OF_OP, NULL, &Contexts[ContextsCount], HandleOneOf);
```

Comment:
When the amount of arguments is too long, it is preferable to split them one argument per line, like this with optional explanation if necessary:

```C
DoForEachOpCode (
  Contexts[ContextsCount].firstIfrHeader,
  EFI_IFR_ONE_OF_OP,
  NULL,                               ///< NULL blahblahblah
  &Contexts[ContextsCount],
  HandleOneOf
  );
```